File: 3106-lexicographically-smallest-string-after-operations-with-constraint/3106-lexicographically-smallest-string-after-operations-with-constraint.cpp
class Solution {
 public:
  string getSmallestString(string s, int k) {
    string ans = s;

    for (char& c : ans) {
      if (k == 0)
        break;
      const int distToA = min(c - 'a', 'z' - c + 1);
      if (k >= distToA) {
        k -= distToA;
        c = 'a';
      } else {
        c -= k;
        k = 0;
      }
    }

    return ans;
  }
};

File: 3106-lexicographically-smallest-string-after-operations-with-constraint/3106-lexicographically-smallest-string-after-operations-with-constraint.py
class Solution:
  def getSmallestString(self, s: str, k: int) -> str:
    ans = list(s)

    for i, c in enumerate(s):
      if k == 0:
        break
      distToA = min(ord(c) - ord('a'), ord('z') - ord(c) + 1)
      if k >= distToA:
        k -= distToA
        ans[i] = 'a'
      else:
        ans[i] = chr(ord(c) - k)
        k = 0

    return ''.join(ans)